modules/core/src/main/java/org/apache/ignite/internal/processors/performancestatistics/FilePerformanceStatisticsWriter.java

```diff
@@ -0,0 +1,517 @@
+import static org.apache.ignite.internal.processors.performancestatistics.OperationType.TASK;
+import static org.apache.ignite.internal.processors.performancestatistics.OperationType.TX_COMMIT;
+import static org.apache.ignite.internal.processors.performancestatistics.OperationType.TX_ROLLBACK;
+
+/**
+ * Performance statistics collector based on logging to a file.
+ * <p>
+ * Each node collects statistics to a file placed under {@link #PERFORMANCE_STAT_DIR}.
+ * <p>
+ * <b>Note:</b> Start again will erase previous performance statistics files.
+ * <p>
+ * To iterate over records use {@link FilePerformanceStatisticsReader}.
+ */
+public class FilePerformanceStatisticsWriter {
+    /** Default maximum file size in bytes. Performance statistics will be stopped when the size exceeded. */
+    public static final long DFLT_FILE_MAX_SIZE = 32 * U.GB;
+
+    /** Default off heap buffer size in bytes. */
+    public static final int DFLT_BUFFER_SIZE = (int)(32 * U.MB);
+
+    /** Default minimal batch size to flush in bytes. */
+    public static final int DFLT_FLUSH_SIZE = (int)(8 * U.MB);
+
+    /** Directory to store performance statistics files. Placed under Ignite work directory. */
+    public static final String PERFORMANCE_STAT_DIR = "performanceStatistics";
+
+    /** Factory to provide I/O interface. */
+    private final FileIOFactory fileIoFactory = new RandomAccessFileIOFactory();
+
+    /** Performance statistics enabled flag. */
+    private volatile boolean enabled;
+
+    /** Performance statistics file writer worker. */
+    @Nullable private volatile FileWriter fileWriter;
+
+    /** Performance statistics file I/O. */
+    @Nullable private volatile FileIO fileIo;
+
+    /** File write buffer. */
+    @Nullable private volatile SegmentedRingByteBuffer ringByteBuffer;
+
+    /** Size of ready for flushing bytes. */
+    private final AtomicInteger readyForFlushSize = new AtomicInteger();
+
+    /** {@code True} if the small buffer warning message logged. */
+    private final AtomicBoolean smallBufLogged = new AtomicBoolean();
+
+    /** {@code True} if worker stopped due to maximum file size reached. */
+    private final AtomicBoolean stopByMaxSize = new AtomicBoolean();
+
+    /** Hashcodes of cached strings. */
+    private final ConcurrentSkipListSet<Integer> cachedStrings = new ConcurrentSkipListSet<>();
+
+    /** Kernal context. */
+    private final GridKernalContext ctx;
+
+    /** Logger. */
+    private final IgniteLogger log;
+
+    /** @param ctx Kernal context. */
+    public FilePerformanceStatisticsWriter(GridKernalContext ctx) {
+        log = ctx.log(getClass());
+
+        this.ctx = ctx;
+    }
+
+    /** Starts collecting performance statistics. */
+    public void start() {
+        synchronized (this) {
+            if (enabled)
+                return;
+
+            enabled = true;
+
+            try {
+                File file = statisticsFile(ctx);
+
+                U.delete(file);
```

Comment:
Let's append to the file if it exists.
It seems all we need to do is remove this line.

We need a ticket to provide `clear` command, also.

Reply:
The previous file can be closed during writing records. So the last record can be written not fully. It may lead to the case when the file can not be read.